wgpu-core/src/resource.rs

```diff
@@ -29,6 +29,13 @@ pub enum BufferMapAsyncStatus {
     ContextLost,
 }
 
+#[derive(Debug)]
+pub enum BufferMapState {
+  Waiting,
```

Comment:
Can we have `Waiting(BufferPendingMapping)` so that `pending_mapping` is not needed any more?

Reply:
Yeah we can have that with just a small change- `Waiting(Option<BufferPendingMapping>)`- so that for transfering the mapping here-https://github.com/gfx-rs/wgpu/blob/05ba7a50b4645f6ac6c65edbae295a31c0d7ebef/wgpu-core/src/device/life.rs#L555
I can leave it to `Waiting(None)`

Reply:
in this code we expect it to be `Waiting` and we are switching it to `Active`, so we don't need the `Option<>` at all. We can just do
```rust
let mapping = match mem::replace(&mut buffer.mapping, BufferMapState::Active) {
  BufferMapState::Waiting(pending) => pending,
  _ => panic!("...."),
};
```

Reply:
This will then cause panic in `map_buffer` when we check for `Active` state there.

Reply:
I'm sorry, could you clarify? When would `BufferMapState::Waiting(None)` be useful?

Reply:
Setting `map_state` to `Active` before calling `map_buffer()` will cause a panic when the check is done in `map_buffer()` 
```
if let resource::BufferMapState::Active = buffer.map_state {
 log::error!("Buffer already mapped");
```
(I have changed it to panic as you suggested above; yet to push the changes)

 `Waiting(None)` will just denote that the actual mapping process has started in an async buffer map call.

Reply:
But we are talking about a place where we are handling `ready_to_map`. Calling `map_buffer()` from the user point of view is indeed invalid, until they `unmap_buffer` first.

Reply:
We are calling `map_buffer()` in `handle_mapping()` itself-
```
let mapping = match buffer.map_state {
    resource::BufferMapState::Waiting(ref mut pending_mapping) => pending_mapping.take().unwrap(),   //In case I use Waiting(Option<..>)
    _ => panic!("No pending mapping."),
};
let result = match mapping.op {
    resource::BufferMapOperation::Read { .. } => {
        super::map_buffer(raw, buffer, mapping.sub_range, super::HostMap::Read)
    }
    resource::BufferMapOperation::Write { .. } => {
        super::map_buffer(raw, buffer, mapping.sub_range, super::HostMap::Write)
    }
};
```

Reply:
then we could assume in `map_buffer` that the state is already `Active`? Technically, it's the other part of code that is responsible for the states to be valid and transitioned. Since everything goes through `ready_to_map`, it's ok to switch the states there.

Reply:
Ok, I'll just remove the check in `map_buffer()`. I hope we don't require a check in [`device_set_buffer_sub_data()`](https://github.com/gfx-rs/wgpu/blob/05ba7a50b4645f6ac6c65edbae295a31c0d7ebef/wgpu-core/src/device/mod.rs#L583) and [`device_set_buffer_sub_data()`](https://github.com/gfx-rs/wgpu/blob/05ba7a50b4645f6ac6c65edbae295a31c0d7ebef/wgpu-core/src/device/mod.rs#L618)

Reply:
these methods will need to go soon, they are wrong :)